### PR TITLE
gemsafeV1 ISO7816 PIN support and PTEID improvements

### PR DESCRIPTION
**Pull request by nunojpg**

### CODE REVIEW

**Comment by frankmorgner** [comment]
bea1850d05cc9ea955e3fbc412c3285da2caa30b looks OK.

Ideally, f825a9bafd3eceee072a5173a5b263af793eef06 should be split up into multiple commits, one commit for each change.

Problem for both commits is, that I don't know who could confirm your changes with an actual card...

**Comment by nunojpg** [comment]
@l1k are you able to test the iso7816 PIN support with bea1850, with the European Patent Office smart card?

**Comment by nunojpg** [comment]
I've splited the commits and I've pinged gemsafe and pteid developers. If there is no NACK I will ping again for this to be committed.

File: src/libopensc/pkcs15-pteid.c [hunk 1]
```diff
@@ -147,20 +147,24 @@ static int sc_pkcs15emu_pteid_init(sc_pkcs15_card_t * p15card)
 		static const char *pteid_pin_names[3] = {
 			"Auth PIN",
 			"Sign PIN",
-			"Address PIN"
-		};
+			"Address PIN"};
```
Comment (l1k) [inline src/libopensc/pkcs15-pteid.c]: I'd leave the curly bracket where it is.

File: src/libopensc/pkcs15-pteid.c [hunk 2]
```diff
@@ -106,22 +107,24 @@ static int sc_pkcs15emu_pteid_init(sc_pkcs15_card_t * p15card)
...
+			"SIGNATURE SUB CA",
+			"AUTHENTICATION SUB CA",
+			"ROOT CA"
 		};
 		/* X.509 Certificate Paths */
-		static const char *pteid_cert_paths[4] = {
+		static const char *pteid_cert_paths[5] = {
 			"3f005f00ef09", /* Authentication Certificate path */
-			"3f005f00ef08", /* Digital Signature Certificate path */
+			"3f005f00ef08", /* Digital Signature Certificate path */
```
Comment (l1k) [inline src/libopensc/pkcs15-pteid.c]: Please remove the trailing whitespace.

File: src/libopensc/pkcs15-pteid.c [hunk 3]
```diff
@@ -106,22 +107,24 @@ static int sc_pkcs15emu_pteid_init(sc_pkcs15_card_t * p15card)
...
-				"SIGNATURE SUB CA",
-				"AUTHENTICATION SUB CA"
+	for (i = 0; i < 5; i++) {
+		static const char *pteid_cert_names[5] = {
+			"AUTHENTICATION CERTIFICATE",
+			"SIGNATURE CERTIFICATE",
+			"SIGNATURE SUB CA",
+			"AUTHENTICATION SUB CA",
+			"ROOT CA"
 		};
```
Comment (l1k) [inline src/libopensc/pkcs15-pteid.c]: I'd leave the indentation unchanged so as not to obscure the "meat" of your patch.

File: src/libopensc/pkcs15-pteid.c [hunk 4]
```diff
@@ -106,22 +107,24 @@ static int sc_pkcs15emu_pteid_init(sc_pkcs15_card_t * p15card)
...
 		/* X.509 Certificate Paths */
-		static const char *pteid_cert_paths[4] = {
+		static const char *pteid_cert_paths[5] = {
 			"3f005f00ef09", /* Authentication Certificate path */
-			"3f005f00ef08", /* Digital Signature Certificate path */
+			"3f005f00ef08", /* Digital Signature Certificate path */			
 			"3f005f00ef0f", /* Signature sub CA path */
-			"3f005f00ef10"	/* Authentication sub CA path */
+			"3f005f00ef10",	/* Authentication sub CA path */
+			"3f005f00ef11"  /* Root CA path */
```
Comment (l1k) [inline src/libopensc/pkcs15-pteid.c]: Please add a comma after the string literal.

Comment (nunojpg) [inline src/libopensc/pkcs15-pteid.c]: All done beside this. I can't get what you mean... Can you check again if any comma is missing?

**Comment by l1k** [comment]
I've tested nunojpg/OpenSC@f825a9bafd3eceee072a5173a5b263af793eef06 with my European Patent Office GemSAFE V1 and V3 cards and PIN authentication still works fine. I've also briefly reviewed your changes and left some line comments on the commits. These all concern style issues. I cannot say anything about the functional changes as I don't have  PTEID card. Generally I'd put one commit at the front which contains all the indentation changes, followed by commits containing the functional changes. By not mixing these up, you avoid obscuring the "meat" of your functional changes. Finally, it would be good if the commit messages explained more verbosely the "why" behind the functional changes. What existing problem do they solve?

File: src/libopensc/pkcs15-pteid.c (hunk 4 above)
Comment (l1k) [inline src/libopensc/pkcs15-pteid.c]: If you add a comma after the "3f005f00ef11" string literal, when someone needs to add another value, they don't have to touch that line. (As you had to with "3f005f00ef10".) Sorry for not making this clearer.

Comment (nunojpg) [inline src/libopensc/pkcs15-pteid.c]: I had NOOO idea that was usual, not even legal syntax. I see the rational for source control and I really like the idea. Thanks!

**Pull request closed without merging**